src/DruidDataSource.ts

```diff
@@ -1,34 +1,329 @@
 import { DataSourceInstanceSettings, MetricFindValue, ScopedVars } from '@grafana/data';
-import { DataSourceWithBackend, getTemplateSrv } from '@grafana/runtime';
-import { DruidSettings, DruidQuery } from './types';
+import { DataSourceWithBackend, getTemplateSrv, logWarning } from '@grafana/runtime';
+import { SQL, type SqlExpression } from './util/SQL';
+import type { DruidSettings, DruidQuery, AdhocVariableModel, AdhocFilter } from './types';
 
 const druidVariableRegex = /\"\[\[(\w+)(?::druid:(\w+))?\]\]\"|\"\${(\w+)(?::druid:(\w+))?}\"/g;
 
 export class DruidDataSource extends DataSourceWithBackend<DruidQuery, DruidSettings> {
   settingsData: DruidSettings;
+
   constructor(instanceSettings: DataSourceInstanceSettings<DruidSettings>) {
     super(instanceSettings);
-    this.settingsData = instanceSettings.jsonData;
+
+    // This proxy is needed for the JS to actually adheer to the current TS type
+    this.settingsData = new Proxy(instanceSettings.jsonData, {
+      get(_, key: string) {
+        return Object.fromEntries(
+          Object.entries(instanceSettings.jsonData)
+            .filter((it) => it[0].startsWith(key))
+            .map((it) => [it[0].substring(`${key}.`.length), it[1]])
+        );
+      },
+    });
   }
+
+  /**
+   * Provides autocompletion for adhoc filter values given the chosen key
+   */
+  async getTagValues(options: { key: string }): Promise<MetricFindValue[]> {
+    if (this.settingsData.adhoc?.shouldNotAutocompleteValue) {
+      return [];
+    }
+
+    const tableNames = (
+      await this._postSqlQuery(
+        `SELECT "TABLE_NAME" FROM INFORMATION_SCHEMA.COLUMNS WHERE "TABLE_SCHEMA" = 'druid' AND "COLUMN_NAME" = '${options.key}'`
+      )
+    ).map((it) => it.value);
+
+    const completions = (
+      await Promise.all(
+        tableNames.map(async (tableName) =>
+          this._postSqlQuery(
+            this.settingsData.adhoc?.shouldNotLimitAutocompleteValue
+              ? `SELECT DISTINCT "${options.key}" FROM ${tableName}`
+              : `SELECT "${options.key}" FROM ${tableName} GROUP BY "${options.key}" ORDER BY COUNT("${options.key}") DESC LIMIT 1000`
```

Comment:
Just text templating in values into SQL queries leaves them vulnerable to SQL injections attacks. Prepared statements ought to be preferred if at all possible, even if there are attempts to sanitise the templated values elsewhere.